### PR TITLE
[doc][build] Script to update cache's file timestamp and source directory

### PR DESCRIPTION
**Pull request by khluu**
- Cache's environment pickle file contains source directory that is from the machine that built the doc originally. Source dir is modified to match with the current system.
- Update source file timestamp so that they don't get marked outdated and automatically rebuilt.
- Update timestamp of files other than sourcefile so they don't get marked recently added and automatically rebuilt.

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
(for @can-anyscale to review)

File: ci/ray_ci/doc/BUILD.bazel [hunk 1]
```diff
@@ -14,6 +14,14 @@ py_binary(
     exec_compatible_with = ["//:hermetic_python"],
 )
 
+py_binary(
+    name = "update_cache_env",
```
Comment (can-anyscale) [inline ci/ray_ci/doc/BUILD.bazel:18]: maybe add the `cmd_` prefix; those files with `cmd_` in the prefix are excluded from `py_library`, so they are not double included (both in the library and in the binary)

File: ci/ray_ci/doc/update_cache_env.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import time
+from typing import List
+from datetime import datetime, timezone
+import click
+
+
+def list_pending_files(ray_dir: str) -> List[str]:
+    """List all files that are added/modified in git repo."""
+    pending_files = []
+    with open(f"{ray_dir}/pending_files.txt", "r") as f:
```
Comment (can-anyscale) [inline ci/ray_ci/doc/update_cache_env.py:13]: where is this txt file coming from? also maybe put the `pending_files.txt` into a constant

File: ci/ray_ci/doc/update_cache_env.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def update_environment_pickle(ray_dir: str, pending_files: List[str]) -> None:
+    """
+    Update the environment pickle file with
+    new source and doctree directory, and modify source file timestamps.
+    """
+    ray_doc_dir = os.path.join(ray_dir, "doc")
+    with open(
+        os.path.join(ray_doc_dir, "_build/doctrees/environment.pickle"), "rb"
```
Comment (can-anyscale) [inline ci/ray_ci/doc/update_cache_env.py:26]: the `_build/doctrees/environment.pickle` is defined as a constant elsewhere (in build_cache i think), maybe just import it here

File: ci/ray_ci/doc/update_cache_env.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    ) as f:
+        pickle.dump(env, f, pickle.HIGHEST_PROTOCOL)
+
+
+def update_file_timestamp(ray_dir: str) -> None:
+    """
+    Update files other than source files to
+    an old timestamp to avoid rebuilding them.
+    """
+    ray_doc_dir: str = os.path.join(ray_dir, "doc")
```
Comment (can-anyscale) [inline ci/ray_ci/doc/update_cache_env.py:59]: remove the `:  str` perhaps, to be consistent with the code base (just use type inference)

File: ci/ray_ci/doc/update_cache_env.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    new_timestamp = datetime(2000, 7, 22, 0, 0, tzinfo=timezone.utc).timestamp()
+
+    # Update all template files timestamp
+    directory = f"{ray_doc_dir}/source/_templates"
```
Comment (can-anyscale) [inline ci/ray_ci/doc/update_cache_env.py:63]: I don't completely understand why we need to change the timestamp on this directory yet, let me find your offline

**Review by can-anyscale — COMMENTED** [review]
do you want to add some unit tests for the changing timestamp logic

File: ci/ray_ci/doc/update_cache_env.py (hunk 4 above)
Comment (khluu) [reply ci/ray_ci/doc/update_cache_env.py]: done

File: ci/ray_ci/doc/update_cache_env.py (hunk 3 above)
Comment (khluu) [reply ci/ray_ci/doc/update_cache_env.py]: done

File: ci/ray_ci/doc/update_cache_env.py (hunk 2 above)
Comment (khluu) [reply ci/ray_ci/doc/update_cache_env.py]: done! It's coming from the load doc cache step (in the other PR)

File: ci/ray_ci/doc/BUILD.bazel (hunk 1 above)
Comment (khluu) [reply ci/ray_ci/doc/BUILD.bazel]: done

File: ci/ray_ci/doc/cmd_update_cache_env.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import time
+from typing import List
+from datetime import datetime, timezone
+import click
+from ci.ray_ci.doc.build_cache import ENVIRONMENT_PICKLE
+
+PENDING_FILES_PATH = "pending_files.txt"
+
+
+def list_pending_files(ray_dir: str) -> List[str]:
```
Comment (can-anyscale) [inline ci/ray_ci/doc/cmd_update_cache_env.py:13]: nit: `list_added_or_modified_files`, same as what you use in the docstrings

**Review by can-anyscale — APPROVED** [review]

**Pull request merged**